# Alliance of Genome Resources VEP pipelines

A collection of pipelines for tasks associated with VEP annotation of variations.  The code contained within this repository is designed to run on the EBI's compute farm using the EnsEMBL Hive pipeline manager.

See the README files in each folder for details of the individual pipelines.


## VepProteinFunction

- Generates databases of SIFT and PolyPhen scores and predictions for the MOD species.  
- Databases are accessed by a VEP plugin to add SIFT/PolyPhen annotation to the VEP output.
- Translated protein sequences are constructed from FASTA, GFF, and (optionally) BAM files.
- Serialized prediction matrices are stored containing all possible amino acid substitutions for each sequence, accessed by the hex md5 of the sequence.
- FULL mode generates the database from scratch.
- UPDATE mode checks for existing prediction matrices, or sequences for which there were valid reasons for being unable to generate protein function annotations, and updates the database for new sequences only.


## ModVep

- Runs VEP on MOD high throughput variation VCF files
- Splits input files, runs VEP in parallel, then combines the output.
- Uses MOD GFF, FASTA, and (optionally) BAM files to construct translated protein sequences.
- Retrieves SIFT and PolyPhen annotations from databases generated by the VepProteinFunction pipeline.


## Running for AGR

The script `run_agr_vep_pipelines.pl` can be used to run both the pipelines above, in addition to running the VEP for phenotypic variants.

The following environmental variables need to be set:

- `AGR_VEP_BASE_DIR` - the directory in which results will be stored
- `AGR_VEP_REPO_DIR` - the path to the directory containing this repository
- `AGR_RELEASE` - the release number of the AGR release, e.g. 8.2.0
- `TOKEN` - the authorisation token used by AGR Data Quartermasters when submitting data to the FMS
- `DOWNLOAD_DATE` - the date on which the pipeline was initiated in the format YYYYMMDD
- `HIVE_ROOT_DIR` - path to clone of [Ensembl Hive GitHub resitory](https://github.com/Ensembl/ensembl-hive)
- `HTP_VEP_WORKING_DIR` - path to directory for intermediate working files created during running of ModVep pipeline
- `PATH_PRED_WORKING_DIR` - path to directory for intermediate working files created during running of VepProteinFunction pipeline
- `PATH_PRED_SQL_DUMP_DIR` - directory in which to store backup dumps of pathogenicity prediction SQL databases
- `PATH_PRED_DB_PREFIX` - prefix for MOD pathogenicity prediction SQL database names, e.g. agr_pathogenicity_predictions_
- `VEP_DIR` - path to clone of [Ensembl VEP GitHub resitory](https://github.com/Ensembl/ensembl-vep)
- `VEP_DBHOST` - host name of MySQL server with eHive and pathogenicity prediction databases
- `VEP_DBUSER` - username for MySQL server with eHive and pathogenicity prediction databases
- `VEP_DBPORT` - port number of MySQL server with eHive and pathogenicity prediction databases
- `SIFT_DIR` - directory containing SIFT installation
- `PPH_DIR` - directory containing PolyPhen-2 installation
- `PPH_CONF_DIR` - path to `VepProteinFunction/polyphen_config_files` folder in this repository
- `NCBI_DIR` - path to directory where psiblast and makeblastdb are installed
- `BLAST_DB` - path to directory containing UniRef90 BLAST database files
- `PPH_BLAST_DB` - path to directory containing UniRef100 BLAST database files
- `UNIPROT_DBS` - path to directory containing uniprot species-specific BLAST protein database files